pkg/issuer/acme/dns/rfc2136/rfc2136.go

```diff
@@ -0,0 +1,188 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package rfc2136 implements a DNS provider for solving the DNS-01 challenge
+// using the rfc2136 dynamic update.
+// This code was adapted from lego:
+// 	  https://github.com/xenolf/lego
+
+package rfc2136
+
+import (
+	"fmt"
+	"net"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/jetstack/cert-manager/pkg/issuer/acme/dns/util"
+	"github.com/miekg/dns"
+)
+
+// SupportedAlgorithms should refer to https://tools.ietf.org/html/rfc4635#section-2
+// but miekd/dns supports only the ones below
+var SupportedAlgorithms = map[string]string{
+	"HMACMD5":    dns.HmacMD5,
+	"HMACSHA1":   dns.HmacSHA1,
+	"HMACSHA256": dns.HmacSHA256,
+	"HMACSHA512": dns.HmacSHA512,
+}
+
+// DNSProvider is an implementation of the acme.ChallengeProvider interface that
+// uses dynamic DNS updates (RFC 2136) to create TXT records on a nameserver.
+type DNSProvider struct {
+	nameserver    string
+	tsigAlgorithm string
+	tsigKeyName   string
+	tsigSecret    string
+}
+
+// NewDNSProvider returns a DNSProvider instance configured for rfc2136
+// dynamic update. Configured with environment variables:
+// RFC2136_NAMESERVER: Network address in the form "host" or "host:port".
+// RFC2136_TSIG_ALGORITHM: Defaults to hmac-md5.sig-alg.reg.int. (HMAC-MD5).
+// See https://github.com/miekg/dns/blob/master/tsig.go for supported values.
+// RFC2136_TSIG_KEY: Name of the secret key as defined in DNS server configuration.
+// RFC2136_TSIG_SECRET: Secret key payload.
+// To disable TSIG authentication, leave the RFC2136_TSIG* variables unset.
+func NewDNSProvider() (*DNSProvider, error) {
+	nameserver := os.Getenv("RFC2136_NAMESERVER")
+	tsigAlgorithm := os.Getenv("RFC2136_TSIG_ALGORITHM")
+	tsigKeyName := os.Getenv("RFC2136_TSIG_KEY_NAME")
+	tsigSecret := os.Getenv("RFC2136_TSIG_SECRET")
+	return NewDNSProviderCredentials(nameserver, tsigAlgorithm, tsigKeyName, tsigSecret)
+}
+
+// NewDNSProviderCredentials uses the supplied credentials to return a
+// DNSProvider instance configured for rfc2136 dynamic update. To disable TSIG
+// authentication, leave the TSIG parameters as empty strings.
+// nameserver must be a network address in the form "IP" or "IP:port".
+func NewDNSProviderCredentials(nameserver, tsigAlgorithm, tsigKeyName, tsigSecret string) (*DNSProvider, error) {
+
+	if nameserver == "" {
+		return nil, fmt.Errorf("RFC2136 nameserver missing")
+	}
+
+	// Append the default DNS port if none is specified.
+	if _, _, err := net.SplitHostPort(nameserver); err != nil {
+		if strings.Contains(err.Error(), "missing port") {
+			host := nameserver
+			if ipaddr := net.ParseIP(host); ipaddr != nil {
+				nameserver = net.JoinHostPort(host, "53")
+			} else {
+				return nil, fmt.Errorf("RFC2136 nameserver must be a valid IP Address, not %v", nameserver)
```

Comment:
How come the nameserver must be an IP address and not a hostname *iff* the nameserver does not contain an explicit port declaration?

From what I can see, this means:

* `nameserver.com:53` is valid
* `8.8.8.8:53` is valid
* `8.8.8.8` is valid
- `nameserver.com` is *not* valid

It seems like we could also make `nameserver.com` default to `nameserver.com:53`?

Reply:
Additionally, this validation should occur in the API validation methods too 😄

Reply:
I've fixed the logic and put other tests, sorry and thanks for catching that!
I've also added tests for invalid dns server in the API.